File: src/Benchmark.java
import java.util.HashMap;
import java.util.Map;
import java.time.Instant;

class Xorshift {
    private long state0, state1;

    // Constructor initializes state using the current timestamp
    public Xorshift() {
        long timestamp = Instant.now().toEpochMilli();
        this.state0 = timestamp ^ 0xDEADBEEF;
        this.state1 = (timestamp << 21) ^ 0x95419C24A637B12FL;
    }

    // Xorshift+ algorithm for generating 64-bit random numbers
    public long next() {
        long s1 = state0;
        long s0 = state1;
        state0 = s0;

        s1 ^= (s1 << 23);
        s1 ^= (s1 >>> 18);
        s1 ^= s0;
        s1 ^= (s0 >>> 5);

        state1 = s1;
        return state1 + s0;
    }
}

public class Benchmark {
    private static final int STRING_COUNT = 10_000;
    private static final int STRING_LENGTH = 8;
    private static final char[] CHARSET = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".toCharArray();

    // Function to generate an 8-character random string
    private static String generateRandomString(Xorshift rng) {
        StringBuilder sb = new StringBuilder(STRING_LENGTH);
        for (int i = 0; i < STRING_LENGTH; i++) {
            int randIndex = Math.abs((int) (rng.next() % CHARSET.length));  // Fix: Ensure positive index
            sb.append(CHARSET[randIndex]);
        }
        return sb.toString();
    }

    // Function to generate `count` random strings and track their occurrences
    private static Map<String, Integer> generateRandomStrings(int count) {
        Xorshift rng = new Xorshift();
        Map<String, Integer> stringCounts = new HashMap<>();

        for (int i = 0; i < count; i++) {
            String randomString = generateRandomString(rng);
            stringCounts.put(randomString, stringCounts.getOrDefault(randomString, 0) + 1);
        }

        return stringCounts;
    }

    // Function to measure execution time
    private static long measureExecutionTime() {
        long start = System.nanoTime();
        generateRandomStrings(STRING_COUNT);
        long end = System.nanoTime();
        return (end - start) / 1_000_000; // Convert to milliseconds
    }

    public static void main(String[] args) {
        long totalTime = 0;
        long[] times = new long[3];

        for (int i = 0; i < 3; i++) {
            long execTime = measureExecutionTime();
            totalTime += execTime;
            System.out.print(execTime + "\t");
            times[i] = execTime;
        }

        double averageTime = totalTime / 3.0;
        System.out.printf("%.2f\n", averageTime);
    }
}

File: src/benchmark.go
package main

import (
	"fmt"
	"time"
)

// Xorshift+ struct
type Xorshift struct {
	state0 uint64
	state1 uint64
}

// Initialize Xorshift+ using the current timestamp
func NewXorshift() *Xorshift {
	timestamp := uint64(time.Now().UnixMilli())
	return &Xorshift{
		state0: timestamp ^ 0xDEADBEEF,
		state1: (timestamp << 21) ^ 0x95419C24A637B12F,
	}
}

// Xorshift+ PRNG function
func (x *Xorshift) Next() uint64 {
	s1 := x.state0
	s0 := x.state1
	x.state0 = s0

	s1 ^= s1 << 23
	s1 ^= s1 >> 18
	s1 ^= s0
	s1 ^= s0 >> 5

	x.state1 = s1
	return x.state1 + s0
}

// Generate an 8-character random string
func generateRandomString(rng *Xorshift) string {
	const charset = "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
	result := make([]byte, 8)

	for i := range result {
		randIndex := int(rng.Next() % uint64(len(charset)))
		result[i] = charset[randIndex]
	}

	return string(result)
}

// Generate `count` random strings and track occurrences
func generateRandomStrings(count int) map[string]int {
	rng := NewXorshift()
	stringCounts := make(map[string]int)

	for i := 0; i < count; i++ {
		randomString := generateRandomString(rng)
		stringCounts[randomString]++
	}

	return stringCounts
}

// Measure execution time in milliseconds
func measureExecutionTime() float64 {
	start := time.Now()
	generateRandomStrings(10_000)
	duration := time.Since(start).Seconds() * 1000 // Convert to milliseconds
	return duration
}

// Main function to run the test 3 times
func main() {
	var totalTime float64
	times := make([]float64, 3)

	for i := 0; i < 3; i++ {
		execTime := measureExecutionTime()
		totalTime += execTime
		fmt.Printf("%.2f\t", execTime)
		times[i] = execTime
	}

	averageTime := totalTime / 3
	fmt.Printf("%.2f\n", averageTime)
}
